test/MUXControlsTestApp/VisualTreeTestBase.cs

```diff
@@ -0,0 +1,360 @@
+                Verify.Fail(error);
+            }
+        }
+
+        // Avoid too much logs, and only log message when !flag.
+        private void CheckTrue(bool flag, string message)
+        {           
+            if (!flag)
+            {
+                Verify.Fail(message);
+            }
+        }
+
+        private bool TestContextContainsKey(string key)
+        {
+#if USING_TAEF
+           return TestContext.Properties.Contains(key);
+#else
+            return TestContext.Properties.ContainsKey(key);
+#endif
+        }
+    }
+
+    class MasterFileStorage
+    {
+        private StorageFolder _storage;
+        public string StorageLocation { get; private set; }
+        public string ExpectedMasterFileName { get; private set; }
+        public string BestMatchedMasterFileName { get; private set; }
+        public MasterFileStorage(bool useLocalStorage, string masterFileNamePrefix, string theme)
+        {
+            _storage = useLocalStorage ? ApplicationData.Current.LocalFolder : KnownFolders.MusicLibrary;
+            string prefix = String.IsNullOrEmpty(theme) ? masterFileNamePrefix : masterFileNamePrefix + "-" + theme;
+            ExpectedMasterFileName = GetExpectedMasterFileName(prefix);
+            BestMatchedMasterFileName = SearchBestMatchedMasterFileName(prefix);
+            StorageLocation = useLocalStorage ? ApplicationData.Current.LocalFolder.Path : "MusicLibrary";
+        }
+
+        public void LogMasterFile(string fileName, string content)
+        {
+            LogMasterFile(_storage, fileName, content);
+        }
+
+        private string GetExpectedMasterFileName(string fileNamePrefix)
+        {
+            return String.Format("{0}-{1}.xml", fileNamePrefix, PlatformConfiguration.GetCurrentAPIVersion());
+        }
+
+        private string SearchBestMatchedMasterFileName(string fileNamePrefix)
+        {
+            for (ushort version = PlatformConfiguration.GetCurrentAPIVersion(); version >= 2; version--)
+            {
+                string fileName = String.Format("{0}-{1}.xml", fileNamePrefix, version);
+                if (MasterFileStorage.IsMasterFilePresent(fileName))
+                {
+                    return fileName;
+                }
+            }
+            {
+                string fileName = fileNamePrefix + ".xml";
+                if (MasterFileStorage.IsMasterFilePresent(fileName))
+                {
+                    return fileName;
+                }
+            }
+            return null;
+        }
+
+        public static string GetMasterFileContent(string fileName)
+        {
+            return GetMasterFileContentAsync(fileName).Result;
+        }
+
+        public static bool IsMasterFilePresent(string fileName)
+        {
+            try
+            {
+                GetMasterFileContentAsync(fileName).Wait();
```

Comment:
Why do you need a blocking call to access the contents just to check if the file exists? You should use a different API for this.

Reply:
StorageFile is a aync IO, For our test case, we need sync call. I didn't see other API which would be sync and read ms-appx:///master/xx.xml

Reply:
I thought there was a separate API for file exists, but maybe not: https://stackoverflow.com/questions/8626018/how-to-check-if-file-exists-in-a-windows-store-app